### PR TITLE
Fail by default if implicit/naked tasks present

### LINKED ISSUE #3866: Decide on the future of naked tasks.
Should they stay or should they go?

https://github.com/cylc/cylc-admin/pull/111#pullrequestreview-506579884

[edit]

"Naked" tasks are tasks without an explicit runtime definition i.e in this example `bar` is a "naked" task:

```
[scheduling]
  [[graph]]
    R1 = """
      foo
      bar
    """

[runtime]
  [[foo]]
```

Naked tasks should cause configuration errors by default with this behaviour overridden by a new config:

```ini
[scheduling]
  allow implicit = True
```

We have referred to these as dummy or naked-dummy tasks in the past, I think we will go with "implicit" from here on?

[/edit]

### PR DESCRIPTION
**Pull request by MetRonnie**
These changes close #3866

Implicit tasks (task without an (empty) section in `flow.cylc[runtime]`) are now disallowed by default, and will raise an error on validation.

Added a setting `flow.cylc[scheduler]allow implicit tasks` to allow them.

Also removed the `--strict` option from `cylc validate`, as it no longer does anything.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered by existing tests.
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/207
- [x] No dependency changes.

### CODE REVIEW

File: tests/functional/graphing/03-filter/flow.cylc [hunk 1]
```diff
@@ -1,27 +0,0 @@
-[scheduling]
```
Comment (oliver-sanders) [inline tests/functional/graphing/03-filter/flow.cylc:1]: Another good catch :clap:

**Review by oliver-sanders — APPROVED** [review]

**Comment by hjoliver** [comment]
(All functional test batches failing).

**Comment by MetRonnie** [comment]
😫  I think we can call this "tests pass"

image

**Comment by oliver-sanders** [comment]
Looking at the timestamps the last updated as 15 minutes before the timeout so this isn't just a matter of an over-zealous timeout:

```
Thu, 11 Mar 2021 13:27:52 GMT [18:57:52] tests/f/reload/08-cycle.t ...................................... ok    83887 ms ( 0.01 usr  0.01 sys + 97.73 cusr 29.58 csys = 127.33 CPU)
Thu, 11 Mar 2021 13:43:31 GMT Error: The action has timed out.
```

This is the result of `cylc scan --states=all` which gives us an idea of what tests were causing the issue:

```
cylctb-20210311T130819Z-klTO/f/jinja2/11-logging (stopped)
cylctb-20210311T130819Z-klTO/f/restart/28-execution-timeout (running)
```

**Comment by MetRonnie** [comment]
It's passing 🎉 (apart from a seemingly spurious pair of codecov checks)

File: CHANGES.md [hunk 2]
```diff
@@ -70,6 +70,11 @@ queueing logic centralized.
 how workflows are restarted
 ([#4040](https://github.com/cylc/cylc-flow/pull/4040)).
 
+"Implicit"/"naked" tasks (tasks that do not have an explicit definition in
+`flow.cylc[runtime]`) are now disallowed by default
+([#4109](https://github.com/cylc/cylc-flow/pull/4109)). You can allow them by
+setting `flow.cylc[scheduler]allow implicit tasks` to `True`.
+
```
Comment (hjoliver) [inline CHANGES.md:77]: We'll need to tidy the change log a little before release - some entries (like this) added outside of the useful "enhancements" and "fixes" sections.

**Review by hjoliver — APPROVED** [review]
Nice, thanks @MetRonnie :+1:

File: CHANGES.md (hunk 2 above)
Comment (MetRonnie) [reply CHANGES.md:77]: I wasn't sure if it was a "notable" change for putting above the others or not

Comment (hjoliver) [reply CHANGES.md:77]: OK, we haven't done that in the past on cylc-flow. Probably a good idea, we'll just need to add the appropriate "notable changes" heading.

**Pull request merged**